src/concurrency/sync.rs

```diff
@@ -354,15 +366,18 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
                         mutex.clock.clone_from(clock)
                     });
                 }
-                if let Some(thread) = this.machine.sync.mutexes[id].queue.pop_front() {
-                    this.unblock_thread(thread, BlockReason::Mutex(id))?;
+                let thread_id = mutex.queue.pop_front();
+                if thread_id.is_some() {
+                    // We need to drop mutex before unblock_thread because mutex will be borrowed again
+                    // in the unblock callback.
+                    drop(mutex);
+                    this.unblock_thread(thread_id.unwrap(), BlockReason::Mutex)?;
                 }
             }
-            Some(old_lock_count)
-        } else {
-            // Mutex is not locked.
-            None
-        })
+            return interp_ok(Some(old_lock_count));
```

Comment:
Why did you change this to use early-return?

Reply:
That's a spurious change introduced when I try to debug fix the reborrow ICE. It is now fixed.